File: libs/language/src/lib/merge/processors/ExpressionPreserver.ts
import { BlueNode } from '../../model';
import { MergingProcessor } from '../MergingProcessor';
import { isNonNullable } from '@blue-labs/shared-utils';

/**
 * Preserves expression values (strings starting with "${" and ending with "}")
 * preventing them from being overridden by type definitions.
 *
 * This processor must run BEFORE TypeAssigner to be effective.
 */
export class ExpressionPreserver implements MergingProcessor {
  process(target: BlueNode, source: BlueNode): BlueNode {
    const sourceValue = source.getValue();

    // Check if source value is an expression
    if (this.isExpression(sourceValue)) {
      // When source has an expression, ignore everything from target
      // and return a minimal node with just the expression
      const expressionNode = new BlueNode();
      if (sourceValue !== undefined) {
        expressionNode.setValue(sourceValue);
      }

      // Preserve name and description from source
      if (source.getName()) {
        expressionNode.setName(source.getName());
      }
      if (source.getDescription()) {
        expressionNode.setDescription(source.getDescription());
      }

      // Important: Don't copy type, properties, or items
      // This prevents type resolution from adding structure

      return expressionNode;
    }

    return target;
  }

  /**
   * Post-process to ensure expressions aren't overridden by subsequent processors
   */
  postProcess(target: BlueNode, source: BlueNode): BlueNode {
    const sourceValue = source.getValue();

    // If source had an expression, make sure it's still preserved
    if (this.isExpression(sourceValue)) {
      // If target lost the expression value, restore it
      if (target.getValue() !== sourceValue && sourceValue !== undefined) {
        const fixed = target.clone();
        fixed.setValue(sourceValue);
        // Remove any type/structure that might have been added
        fixed.setType(undefined);
        fixed.setProperties(undefined);
        fixed.setItems(undefined);
        return fixed;
      }
    }

    return target;
  }

  private isExpression(value: unknown): boolean {
    return (
      isNonNullable(value) &&
      typeof value === 'string' &&
      value.startsWith('${') &&
      value.endsWith('}')
    );
  }
}

File: references/java-sdk/src/main/java/blue/language/merge/processor/ExpressionPreserver.java
package blue.language.merge.processor;

import blue.language.NodeProvider;
import blue.language.merge.MergingProcessor;
import blue.language.merge.NodeResolver;
import blue.language.model.Node;

import java.util.Map;

/**
 * Preserves expression scalar values so type assignment does not coerce them.
 */
public class ExpressionPreserver implements MergingProcessor {

    @Override
    public void process(Node target, Node source, NodeProvider nodeProvider, NodeResolver nodeResolver) {
        if (!hasExpressionValue(source)) {
            return;
        }
        target.value(source.getValue());
        target.properties((Map<String, Node>) null);
        target.items((java.util.List<Node>) null);
        target.type((Node) null);
    }

    @Override
    public void postProcess(Node target, Node source, NodeProvider nodeProvider, NodeResolver nodeResolver) {
        if (!hasExpressionValue(source)) {
            return;
        }
        if (source.getValue() != null && !source.getValue().equals(target.getValue())) {
            target.value(source.getValue());
        }
    }

    private boolean hasExpressionValue(Node source) {
        if (source == null || !(source.getValue() instanceof String)) {
            return false;
        }
        String text = ((String) source.getValue()).trim();
        return text.startsWith("${") && text.endsWith("}");
    }
}
